Improve reliability when communicating with external systems

jsenko opened the pull request:
Any ideas where else to add timeouts or retries?

EricWittmann commented on app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:
```diff
@@ -129,6 +128,13 @@
+        // From io.apicurio.registry.mt.TenantMetadataService:
+        map.put(NotAuthorizedException.class, HTTP_FORBIDDEN);
+        map.put(ForbiddenException.class, HTTP_FORBIDDEN);
+        map.put(RegistryTenantNotFoundException.class, HTTP_NOT_FOUND);
```
I wonder if we should instead map this to HTTP_FORBIDDEN.  I think returning a 404 if the registry tenant is not found has the side effect of leaking information.  This can be used by an attacker to scan for valid tenant IDs.

WDYT?

EricWittmann replied:
Does the PR from @carlesarnal impact this?  I suspect so.

carlesarnal replied:
Yes, it does.

EricWittmann commented on app/src/main/java/io/apicurio/registry/services/auth/CustomAuthenticationMechanism.java:
```diff
@@ -115,6 +118,8 @@ public void init() {
         }
     }
 
+    @Retry(abortOn = {NotAuthorizedException.class, ForbiddenException.class}) // 3 retries, 200ms jitter
+    @Timeout(3000) // 3000ms
     public Uni<SecurityIdentity> customAuthentication(RoutingContext context, IdentityProviderManager identityProviderManager) {
```
Is this where the Retry should happen?  Or should we be more targeted by wrapping the interactions with external systems in CDI beans so we can put the fault tolerance at a more granular location?  Honest question because I'm not sure.

Which lines of code in this method might fail and need to be retried?  It's definitely line 125/130 (OidcAuth).  Is it also 121/126 and 129/134 and 133/138?  It's hard to tell.

carlesarnal replied:
Agreed, I think we should extract some parts of the code here so we can retry at a more granular level. There are some situations in this method that are worth a retry () but others, IMO, don't.

carlesarnal commented on app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:
```diff
@@ -76,6 +74,8 @@
 import io.apicurio.registry.storage.RuleNotFoundException;
 import io.apicurio.registry.storage.VersionNotFoundException;
 
+import static java.net.HttpURLConnection.*;
```
We don't want this, probably.

carlesarnal commented:
@jsenko any updates on this? I think we should add the retries relatively soon to our code.

EricWittmann commented on app/src/main/java/io/apicurio/registry/services/http/RegistryExceptionMapperService.java:
```diff
@@ -129,6 +128,13 @@
+        map.put(NotAuthorizedException.class, HTTP_FORBIDDEN);
+        map.put(ForbiddenException.class, HTTP_FORBIDDEN);
+        map.put(RegistryTenantNotFoundException.class, HTTP_NOT_FOUND);
+        map.put(TenantNotFoundException.class, HTTP_NOT_FOUND);
```
@carlesarnal wdyt?

carlesarnal replied:
`RegistryTenantNotFoundException` is thrown by the client, and then is wrapped with TenantNotFound. Actually, that makes me think that the first one should not be present on this mapper.

jsenko commented:
hi @carlesarnal I reverted changes to the `CustomAuthenticationMechanism`. I'd like to leave the retries in that area to you, as you will know more. I've fixed the small issues from the other comments, so I think we can merge this if tests pass.

carlesarnal commented:
> hi @carlesarnal I reverted changes to the `CustomAuthenticationMechanism`. I'd like to leave the retries in that area to you, as you will know more. I've fixed the small issues from the other comments, so I think we can merge this if tests pass.

Ok, yes, I will probably refactor that area to break it into smaller pieces. I'll continue once we merge this.

carlesarnal reviewed (approved):
LGTM
